### PR TITLE
doc: add section on not reinventing the wheel

### PR DESCRIPTION
**Pull request by danielolsen**
### Purpose

This has been on my mind for a while. Right now, the coding guide is mostly focused on testing (maybe some of that information can be moved to the Testing Guide?), but there's not very much information on what makes code 'good' from a non-testing perspective, which is where I assume most people will start.

### Time to review

5 minutes to read, longer if people want to suggest improvements (please be my guest!).

### CODE REVIEW

File: source/dev/coding_guide.rst [hunk 1]
```diff
@@ -41,6 +42,27 @@ generating testable code is not just about making testing less troublesome, but
...
+both sides having to focus on duplicative code.
+
+For example, in Python:
++ If you want to operate on arrays or matrices of un-labelled data, you probably want to use NumPy (and/or SciPy).
++ If you want to operate on tables of labelled data, you probably want to use Pandas.
++ If you want to analyze a collection of nodes, connected by links, you probably want to use networkx.
+
+It's not a requirement that you work with third-party libraries, but if what you are manually implementing overlaps
+substantially with an existing mature library (and especially if this is a library that we already use within a project),
+we may suggest that the code be re-written before being merged. Is you are not familiar with this library, this is okay:
```
Comment (rouille) [inline source/dev/coding_guide.rst:61]: Is you are not --> If you are not

**Review by rouille — APPROVED** [review]
I like it

File: source/dev/coding_guide.rst (hunk 1 above)
Comment (danielolsen) [reply source/dev/coding_guide.rst]: Done.

File: source/dev/coding_guide.rst [hunk 2]
```diff
@@ -41,6 +42,27 @@ generating testable code is not just about making testing less troublesome, but
...
++ If you want to operate on arrays or matrices of un-labelled data, you probably want to use NumPy (and/or SciPy).
++ If you want to operate on tables of labelled data, you probably want to use Pandas.
++ If you want to analyze a collection of nodes, connected by links, you probably want to use networkx.
+
+It's not a requirement that you work with third-party libraries, but if what you are manually implementing overlaps
+substantially with an existing mature library (and especially if this is a library that we already use within a project),
+we may suggest that the code be re-written before being merged. If you are not familiar with this library, this is okay:
+if you let us know that you are unfamiliar, we may be able to provide a code snippet or a git branch that
+demonstrates this functionality.
+
```
Comment (BainanXia) [inline source/dev/coding_guide.rst:64]: Maybe we could also add here if anyone finds a more efficient implementation in our existing code, it is very welcomed to point it out or submit a PR to refactor and we will appreciate the contribution.

Comment (rouille) [reply source/dev/coding_guide.rst:64]: Perhaps we should reorganize the contribution guide. We could have a first section that explicitly lists the type of contribution with short examples: bug report, feature request, refactoring, documentation, testing, etc. Then, we can go in the details of git, docstring, code guide, etc. 

I was thinking about doing something similar for the installation guide. Depending on what the user wants to do, specify the type of installation needed.

**Review by BainanXia — APPROVED** [review]
Thanks!

**Pull request merged**